

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="fr" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="fr" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Récupérer les identifiants taxonomiques du même genre entre Kraken et Blast &mdash; Documentation CLINICAL METAGENOMICS 0.0.1</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script src="../_static/translations.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Recherche" href="../search.html" />
    <link rel="next" title="Création d’un base de donnée 16S avec le cluster" href="../cluster/cluster_create_16S_database.sh.html" />
    <link rel="prev" title="Comparer les identifiants taxonomiques entre Kraken et Blast" href="../bash/find_same_id_kraken_blast_bacteria.sh.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> CLINICAL METAGENOMICS
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Shell scripts:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../bash/create_kraken_database.sh.html">Créer les bases de données personnalisées avec Kraken 2</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/classify_set_sequences.sh.html">Classifier les séquences</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/create_blast_database_without_low_complexity.sh.html">Création d’une base de donnée pour blast sans les séquences de faibles compléxités</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/launch_blast_analyse.sh.html">Analyse des séquences ou reads avec l’algorithme de Blast</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/launch_preprocess.sh.html">Lancer le préprocess ou nettoyage des séquences ou reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/recover_reads.sh.html">Récupérer les reads d’intérêts</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bash/find_same_id_kraken_blast_bacteria.sh.html">Comparer les identifiants taxonomiques entre Kraken et Blast</a></li>
</ul>
<p class="caption"><span class="caption-text">Python scripts:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Récupérer les identifiants taxonomiques du même genre entre Kraken et Blast</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#les-parametres-d-entres">Les paramètres d’entrés</a></li>
<li class="toctree-l2"><a class="reference internal" href="#exemple-d-execution">Exemple d’execution</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Cluster</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../cluster/cluster_create_16S_database.sh.html">Création d’un base de donnée 16S avec le cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cluster/cluster_create_2_custom_databases_and_classify.sh.html">Création de plusieurs bases de données et classification</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cluster/cluster_launch_blast.sh.html">Alignement des séquences ou reads avec l’algorithme de blast</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cluster/cluster_launch_preprocess.sh.html">Nettoyage des séquences ou reads avec l’action de preprocess</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cluster/cluster_launch_get_id_and_reads.sh.html">Récupérer l’ID taxonomique et les reads d’intérêts</a></li>
</ul>
<p class="caption"><span class="caption-text">Télécharger les bases de données automatiquement</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../download/download_refseq_species_sequences.sh.html">Télécharger les séquences des espèces virales et bactériennes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../download/download_ete3_ncbi_taxonomy_database.sh.html">Télécharger la taxonomie de ncbi avec ete3toolkit</a></li>
</ul>
<p class="caption"><span class="caption-text">Environnement Conda</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../conda/conda_environnement.html">Environnement Conda</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CLINICAL METAGENOMICS</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Récupérer les identifiants taxonomiques du même genre entre Kraken et Blast</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/python/sort_blasted_seq.py.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="recuperer-les-identifiants-taxonomiques-du-meme-genre-entre-kraken-et-blast">
<h1>Récupérer les identifiants taxonomiques du même genre entre Kraken et Blast<a class="headerlink" href="#recuperer-les-identifiants-taxonomiques-du-meme-genre-entre-kraken-et-blast" title="Lien permanent vers ce titre">¶</a></h1>
<p>Recover sequence with conserved classification between Kraken2 and BLAST. This program allow to separate the sequences : those which have a similar taxonomy at the genus level are gathered in the file « conserved.txt » (output).</p>
<div class="section" id="les-parametres-d-entres">
<h2>Les paramètres d’entrés<a class="headerlink" href="#les-parametres-d-entres" title="Lien permanent vers ce titre">¶</a></h2>
<p>Les 3 paramètres sont :</p>
<blockquote>
<div><ul class="simple">
<li><p>-i : The blast file input e.g *.blast.txt .</p></li>
<li><p>-o : The output file for e.g *_conserved.txt</p></li>
<li><p>-n : The localization of NCBI taxa database.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="exemple-d-execution">
<h2>Exemple d’execution<a class="headerlink" href="#exemple-d-execution" title="Lien permanent vers ce titre">¶</a></h2>
<div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>python ../python/sort_blasted_seq.py <span class="se">\</span>
-i <span class="si">${</span><span class="nv">BLAST_FOLDER</span><span class="si">}</span><span class="nv">$EACH_BLAST_FILE</span> <span class="se">\</span>
-o <span class="si">${</span><span class="nv">BASENAME_FILE</span><span class="si">}</span>conserved.txt <span class="se">\</span>
-n <span class="si">${</span><span class="nv">PATH_NCBI_TAXA</span><span class="si">}</span>taxa.sqlite
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../cluster/cluster_create_16S_database.sh.html" class="btn btn-neutral float-right" title="Création d’un base de donnée 16S avec le cluster" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../bash/find_same_id_kraken_blast_bacteria.sh.html" class="btn btn-neutral float-left" title="Comparer les identifiants taxonomiques entre Kraken et Blast" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Zygnematophyce

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>